extensions/grpc/src/main/java/com/hazelcast/jet/grpc/impl/UnaryService.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2020 Hazelcast Inc.
+ *
+ * Licensed under the Hazelcast Community License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.grpc.impl;
+
+import com.hazelcast.function.BiConsumerEx;
+import com.hazelcast.function.FunctionEx;
+import com.hazelcast.jet.JetException;
+import com.hazelcast.jet.grpc.GrpcService;
+import io.grpc.ManagedChannel;
+import io.grpc.stub.StreamObserver;
+
+import javax.annotation.Nonnull;
+import java.util.concurrent.CompletableFuture;
+
+public final class UnaryService<T, R> implements GrpcService<T, R> {
+
+    private final BiConsumerEx<? super T, ? super StreamObserver<R>> callFn;
+
+    public UnaryService(
+            @Nonnull ManagedChannel channel,
+            @Nonnull FunctionEx<? super ManagedChannel, ? extends BiConsumerEx<T, StreamObserver<R>>> callStubFn
+    ) {
+        callFn = callStubFn.apply(channel);
+    }
+
+    @Override @Nonnull
+    public CompletableFuture<R> call(@Nonnull T input) {
+        Observer<R> o = new Observer<>();
+        callFn.accept(input, o);
+        return o.future;
+    }
+
+    public void destroy() {
+    }
+
+    private static class Observer<O> implements StreamObserver<O> {
+        private final CompletableFuture<O> future;
+
+        Observer() {
+            this.future = new CompletableFuture<>();
+        }
+
+        @Override
+        public void onNext(O value) {
+            if (!future.complete(value)) {
+                throw new JetException("UnaryService.Observer.onNext() called, but its future is already completed");
+            }
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            future.completeExceptionally(GrpcUtil.translateGrpcException(t));
+        }
+
+        @Override
+        public void onCompleted() {
```

Comment:
Yes it gets called. originally we were completing the future here, but it requires a `volatile` because gRPC doesn't guarantee same thread calls onComplete with onNext. I think it's more correct to do it it was previously. We don't want to complete the future if it would fail for some reason between onNext() and onCompleted() @mtopolnik what do you think?

Reply:
I think there's no way gRPC will call `onError` after calling `onNext`. Note that this same interface, `StreamObserver`, serves the need of both unary and streaming responses. Either `onNext` or `onError` will be called, followed by `onCompleted`. OTOH, I'm also pretty sure there's a _happens-before_ between callback invocations and that we could also go back and remove `volatile`. I just find this way more robust.

Reply:
you have to call onNext and then onComplete on the server code, so I think it could happen.